operator/v1/types_ingress.go

```diff
@@ -556,6 +572,47 @@ const (
 	AWSNetworkLoadBalancer AWSLoadBalancerType = "NLB"
 )
 
+// AWSSubnets contains a list of references to AWS subnets by
+// ID or name.
+// +kubebuilder:validation:XValidation:rule=`has(self.ids) && has(self.names) ? size(self.ids + self.names) <= 10 : true`,message="the total number of subnets cannot exceed 10"
+// +kubebuilder:validation:XValidation:rule=`has(self.ids) || has(self.names)`,message="must specify at least 1 subnet name or id"
+type AWSSubnets struct {
+	// ids specifies a list of AWS subnets by subnet ID.
+	// Subnet IDs must start with "subnet-", consist only
+	// of alphanumeric characters, must be exactly 24
+	// characters long, and must be unique.
+	//
+	// +optional
+	// +listType=atomic
+	// +kubebuilder:validation:XValidation:rule=`self.all(x, self.exists_one(y, x == y))`,message="subnet ids cannot contain duplicates"
```

Comment:
Do you need CEL validation for maxItems=10?

Reply:
I don't think we need both `// +kubebuilder:validation:MaxItems=10` and CEL to enforce a max item list of 10 items.

The kubebuilder built-in `MaxItems` will do the job of preventing more than 10 items in the list without CEL. From what I understand, they accomplish the same type of validation, but CEL is much more flexible.

Let me know if I'm am misunderstanding the question.

Reply:
I guess this is the CEL that validates that.  If both IDs and Names have more than 10, I wonder what kind of error message the user sees?

> // +kubebuilder:validation:XValidation:rule=`has(self.ids) && has(self.names) ? size(self.ids + self.names) <= 10 : true`,message="the total number of subnets cannot exceed 10"

Perhaps that CEL should not have the check for `<= 10`, and the error message should be that it has to have at least one subnet?

Reply:
Also, maybe it should be `> 1` and `< 10` ?

Reply:
`< 11` 😄

Reply:
> I guess this is the CEL that validates that. If both IDs and Names have more than 10, I wonder what kind of error message the user sees?

Did you mean 10 combined (like 6+6=11)? Or both individually more than 10 (12+12=24)? I'll describe both.

If both lists are individually > 10, you'll get two error messages:
```
# * spec.endpointPublishingStrategy.loadBalancer.providerParameters.aws.subnets.ids: Too many: 12: must have at most 10 items
# * spec.endpointPublishingStrategy.loadBalancer.providerParameters.aws.subnets.names: Too many: 12: must have at most 10 items
``` 

If the sum of both combined is > 10, but each list is individually < 10:
```
# * spec.endpointPublishingStrategy.loadBalancer.providerParameters.aws.subnets: Invalid value: "object": the total number of subnets cannot exceed 10
```

> Perhaps that CEL should not have the check for <= 10, and the error message should be that it has to have at least one subnet?

The other CEL rule already enforces at least 1 subnet:
```
// +kubebuilder:validation:XValidation:rule=`has(self.ids) || has(self.names)`,message="must specify at least 1 subnet name or id"
```

So are you staying we should remove the CEL that checks for the total <= 10? Then the total possible subnets will be increased to 20, but only when you use all of both Subnets and IDs. You can use both IDs and Names simultaneously, so we want to limit the combination to 10, that's what this CEL validation is solving.

The `// +kubebuilder:validation:MaxItems=10` is redundant, yes, because we already limit the combination of all subnets to 10 via CEL. But it's a requirement because if you don't have it, you'll get an error when trying to apply the CRD:
```
Forbidden: estimated rule cost exceeds budget by factor of more than 100x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared)
```

Hence the need for both CEL and MaxItems.

Reply:
> The `// +kubebuilder:validation:MaxItems=10` is redundant, yes, because we already limit the combination of all subnets to 10 via CEL. But it's a requirement because if you don't have it, you'll get an error when trying to apply the CRD:
> 
> ```
> Forbidden: estimated rule cost exceeds budget by factor of more than 100x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared)
> ```
>
> Hence the need for both CEL and MaxItems.

This might warrant adding a `// +` comment to explain why you have the seemingly redundant `MaxItems=10` rule.

Reply:
> This might warrant adding a // + comment to explain why you have the seemingly redundant MaxItems=10 rule.

Didn't realize you could just add arbitrary comments. Done.